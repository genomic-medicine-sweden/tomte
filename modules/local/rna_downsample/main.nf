process RNA_DOWNSAMPLE {
    tag "$meta.id"
    label 'process_low'

    conda "bioconda::samtools=1.16.1"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/samtools:1.16.1--h6899075_1' :
        'biocontainers/samtools:1.16.1--h6899075_1' }"

    input:
    tuple val(meta), path(bam), path(bai)
    val num_reads

    output:
    tuple val(meta), path("*_downsmapled.bam")                                , emit: bam
    tuple val(meta), path("*_downsmapled.bam.bai")                            , emit: bai
    tuple val(meta), path("*_downsmapled.bam"), path("*_downsmapled.bam.bai") , emit: bam_bai
    path  "versions.yml"                                                      , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"

    """
    fraction=\$(samtools idxstats ${bam} | cut -f3 | awk -v ct="${num_reads}" 'BEGIN {total=0} {total += \$1} END {print ct/total}')
    percent=\$(samtools idxstats ${bam} | cut -f3 | awk -v ct="${num_reads}"00 'BEGIN {total=0} {total += \$1} END {print ct/total}')
    percent=\$(echo \$percent | cut -d. -f1)
    if (( \$(echo "\$percent < 100") ))
    then
        samtools view -@ $task.cpus -b ${bam} -s \$fraction -o ${prefix}_downsmapled.bam
    else
        mv ${bam} ${prefix}_downsmapled.bam
    fi
    samtools index ${prefix}_downsmapled.bam

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        samtools: \$(echo \$(samtools --version 2>&1) | sed 's/^.*samtools //; s/Using.*\$//')
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix    ?: "${meta.id}"
    """
    touch ${prefix}_downsmapled.bam
    touch ${prefix}_downsmapled.bam.bai

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        samtools: \$(echo \$(samtools --version 2>&1) | sed 's/^.*samtools //; s/Using.*\$//')
    END_VERSIONS
    """
}
