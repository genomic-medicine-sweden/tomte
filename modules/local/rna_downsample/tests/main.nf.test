nextflow_process {

    name "Test Process RNA_DOWNSAMPLE"
    script "../main.nf"
    process "RNA_DOWNSAMPLE"
    tag "modules_local"
    tag "modules"
    tag "rna_downsample"

    test("Should run without failures") {

        when {
            process {
                """
                input[0] = Channel.of([
                        [ case:'finequagga', id:'ACC5963A1', sample:'ACC5963A1', strandedness:'reverse', sex:'NA', single_end:false, fq_pairs:1, is_fastq:false ],
                        file("${projectDir}/test_data/ACC5963A1.bam", checkIfExists: true),
                        file("${projectDir}/test_data/ACC5963A1.bam.bai", checkIfExists: true)
                    ])
                input[1] = 20000
                """
            }
        }

        then {
            assert process.success
            assert snapshot(process.out).match()
        }

    }

}
