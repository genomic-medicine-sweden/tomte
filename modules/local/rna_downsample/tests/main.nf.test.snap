{
    "Should run without failures": {
        "content": [
            {
                "0": [
                    [
                        {
                            "case": "finequagga",
                            "id": "ACC5963A1",
                            "sample": "ACC5963A1",
                            "strandedness": "reverse",
                            "sex": "NA",
                            "single_end": false,
                            "fq_pairs": 1,
                            "is_fastq": false
                        },
                        "ACC5963A1_downsmapled.bam:md5,482c74fe04c5ea9e78184711b5d68a84"
                    ]
                ],
                "1": [
                    [
                        {
                            "case": "finequagga",
                            "id": "ACC5963A1",
                            "sample": "ACC5963A1",
                            "strandedness": "reverse",
                            "sex": "NA",
                            "single_end": false,
                            "fq_pairs": 1,
                            "is_fastq": false
                        },
                        "ACC5963A1_downsmapled.bam.bai:md5,5dbc44b47537cde7b94a586cec145481"
                    ]
                ],
                "2": [
                    [
                        {
                            "case": "finequagga",
                            "id": "ACC5963A1",
                            "sample": "ACC5963A1",
                            "strandedness": "reverse",
                            "sex": "NA",
                            "single_end": false,
                            "fq_pairs": 1,
                            "is_fastq": false
                        },
                        "ACC5963A1_downsmapled.bam:md5,482c74fe04c5ea9e78184711b5d68a84",
                        "ACC5963A1_downsmapled.bam.bai:md5,5dbc44b47537cde7b94a586cec145481"
                    ]
                ],
                "3": [
                    "versions.yml:md5,ccdfa6141346e338e096f7c990633462"
                ],
                "bai": [
                    [
                        {
                            "case": "finequagga",
                            "id": "ACC5963A1",
                            "sample": "ACC5963A1",
                            "strandedness": "reverse",
                            "sex": "NA",
                            "single_end": false,
                            "fq_pairs": 1,
                            "is_fastq": false
                        },
                        "ACC5963A1_downsmapled.bam.bai:md5,5dbc44b47537cde7b94a586cec145481"
                    ]
                ],
                "bam": [
                    [
                        {
                            "case": "finequagga",
                            "id": "ACC5963A1",
                            "sample": "ACC5963A1",
                            "strandedness": "reverse",
                            "sex": "NA",
                            "single_end": false,
                            "fq_pairs": 1,
                            "is_fastq": false
                        },
                        "ACC5963A1_downsmapled.bam:md5,482c74fe04c5ea9e78184711b5d68a84"
                    ]
                ],
                "bam_bai": [
                    [
                        {
                            "case": "finequagga",
                            "id": "ACC5963A1",
                            "sample": "ACC5963A1",
                            "strandedness": "reverse",
                            "sex": "NA",
                            "single_end": false,
                            "fq_pairs": 1,
                            "is_fastq": false
                        },
                        "ACC5963A1_downsmapled.bam:md5,482c74fe04c5ea9e78184711b5d68a84",
                        "ACC5963A1_downsmapled.bam.bai:md5,5dbc44b47537cde7b94a586cec145481"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,ccdfa6141346e338e096f7c990633462"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-04-16T10:41:23.151292"
    }
}